scripts/sync_fofb/scan_dcc_links.py

```diff
@@ -0,0 +1,101 @@
+#! /usr/bin/python3
+
+''' Script to scan FOFB DCC FMC/P2P links.
+
+    This script reads DCC link partners PVs, builts a directed graph and saves
+    it on disk using json format.
+
+    Usage:  "./scan_dcc_links.py <JSON FILENAME> <LIST OF CRATES>", where
+
+            <JSON FILENAME>:    output filename
+            <LIST OF CRATES>:   list of crates' indexes separated by spaces
+
+    Author: Guilherme Ricioli
+'''
+
+import collections
+from epics import PV
+from itertools import product
+import json
+import sys
+
+import bpm_epics_ioc_slot_mapping
+
+# crates
+crates = [f"{int(i):02}" for i in sys.argv[2:]]
+
+# logical slots
+rtmlamp_slot = 3
+slots = [rtmlamp_slot, 13, 15, 17, 19]
+
+num_of_gts = 8
+link_partners_pvs_suffix = ["LinkPartnerCH" + f"{gt}" + "-Mon" for gt in range(num_of_gts)]
+
+def consume(iterator):
+	collections.deque(iterator, maxlen=0)
+
+global_pv_list = []
+def create_pv(name):
+	pv = PV(name)
+	global_pv_list.append(pv)
+	return pv
+
+def pv_prefix_gen(slot, crate):
+	pv_prefix = ""
+	if slot == rtmlamp_slot:
+		# board connected to physical slot 2 == RTMLAMP
+		pv_prefix = "IA-" + crate + "RaBPM:BS-FOFBCtrl:"
+	else:
+		key_prefix = "CRATE_" + str(crate) + "_BPM_" + str(slot)
+		pv_prefix += bpm_epics_ioc_slot_mapping.area_prefix_dict[key_prefix + "_PV_AREA_PREFIX"]
+		pv_prefix += bpm_epics_ioc_slot_mapping.device_prefix_dict[key_prefix + "_PV_DEVICE_PREFIX"]
+
+	return pv_prefix
+
+def fofb_ctrl_pv_list_gen(name, slot, crate):
+	pv_prefix = pv_prefix_gen(slot, crate)
+	pv_list = [create_pv(pv_prefix + "DCCP2P" + name)]
+	if slot == rtmlamp_slot:
+		# has additional FOFB_CC core
+		pv_list.append(create_pv(pv_prefix + "DCCFMC" + name))
+
+	return pv_list
+
+def pv_list_get(pv_list, wait=True, check=True):
+	val_list = []
+	for pv in pv_list:
+		val_list.append(pv.get())
+
+	return val_list
+
+pvs_dict = {}
+for key in product(crates, slots):
+	crate, slot = key
+
+	pvs_ids = fofb_ctrl_pv_list_gen("BPMId-RB", slot, crate)
+	pvs_neighbors = []
+	for pv_suffix in link_partners_pvs_suffix:
+		pvs_neighbors.extend(fofb_ctrl_pv_list_gen(pv_suffix, slot, crate))
+
+	pvs_dict[key] = (pvs_ids, pvs_neighbors)
+
+consume((pv.wait_for_connection() for pv in global_pv_list))
+
+dcc_graph = {}
+for key in product(crates, slots):
+	crate, slot = key
+
+	pvs_ids, pvs_neighbors = pvs_dict[key]
+
+	# using DCCP2PBPMId as id
+	id = pv_list_get(pvs_ids)[0]
+	neighbors = pv_list_get(pvs_neighbors)
+	neighbors.sort()
+	neighbors = list(dict.fromkeys(neighbors)) # removing duplicates
```

Comment:
It's not clear to me how this works. If you're dealing with duplicates, could you use a `set` instead?

Reply:
I could, but I want `neighbors` to be ordered. Is there a better way to achieve that?

Reply:
```
neighbors = list(set(neighbors)).sort()
```

What do you think?

Reply:
I would need to break it into

```suggestion
	neighbors = list(set(neighbors)) # removing duplicates
	neighbors.sort()
```
since `sort()` sorts the list in-place. Doesn't seem to optimize anything for me, though it's clearer... I'll change it.